File: Dai-Hentai/Views/GalleryCell/GalleryCell.m
//
//  GalleryCell.m
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2017/2/14.
//  Copyright © 2017年 DaidoujiChen. All rights reserved.
//

#import "GalleryCell.h"
#import "UIAlertController+Block.h"

@interface GalleryCell ()

@property (nonatomic, assign) BOOL isPresenting;
@property (nonatomic, readonly) UIViewController *rootViewController;

@end

@implementation GalleryCell

#pragma mark - Private Instance Method

- (UIViewController *)rootViewController {
    return [UIApplication sharedApplication].keyWindow.rootViewController;
}

- (void)handleLongPress:(UILongPressGestureRecognizer *)longPressGestureRecognizer {
    if (self.galleryImageView.image && !self.isPresenting) {
        self.isPresenting = YES;
        UIActivityViewController *activityViewController = [[UIActivityViewController alloc] initWithActivityItems:@[ self.galleryImageView.image ] applicationActivities:nil];
        
        __weak GalleryCell *weakSelf = self;
        activityViewController.completionWithItemsHandler = ^(UIActivityType activityType, BOOL completed, NSArray *returnedItems, NSError *activityError) {
            weakSelf.isPresenting = NO;
        };
        [self.rootViewController presentViewController:activityViewController animated:YES completion:nil];
    }
}

#pragma mark - Life Cycle

- (id)initWithCoder:(NSCoder *)aDecoder {
    UILongPressGestureRecognizer *longPressGestureRecognizer = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(handleLongPress:)];
    longPressGestureRecognizer.delaysTouchesBegan = YES;
    longPressGestureRecognizer.minimumPressDuration = 1.0f;
    [self addGestureRecognizer:longPressGestureRecognizer];
    self.isPresenting = NO;
    return [super initWithCoder:aDecoder];
}

- (void)prepareForReuse {
    [super prepareForReuse];
    self.galleryImageView.image = nil;
    self.galleryImageView.backgroundColor = [UIColor blackColor];
}

@end

File: Dai-Hentai/Views/GalleryCell/GalleryCell.swift
//
//  GalleryCell.swift
//  Dai-Hentai
//
//  Created by DaidoujiChen on 2018/12/19.
//  Copyright © 2018 DaidoujiChen. All rights reserved.
//

import Foundation
import UIKit

class GalleryCell: UICollectionViewCell {
    
    // MARK: - Property
    
    @IBOutlet weak var imageView: UIImageView!
    
    private var isPresenting = false
    private var rootViewController: UIViewController? {
        return UIApplication.shared.keyWindow?.rootViewController
    }
    
    // MARK: - Life Cycle
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        let longPressGestureRecognizer = UILongPressGestureRecognizer(target: self, action: #selector(handleLongPress(_:)))
        longPressGestureRecognizer.delaysTouchesBegan = true
        longPressGestureRecognizer.minimumPressDuration = 1
        addGestureRecognizer(longPressGestureRecognizer)
    }
    
    override func prepareForReuse() {
        super.prepareForReuse()
        imageView.image = nil
        imageView.backgroundColor = .black
    }
    
    // MARK: - Function
    
    @objc private func handleLongPress(_ gesture: UILongPressGestureRecognizer) {
        guard
            let image = imageView.image,
            !isPresenting else {
            return
        }
        isPresenting = true
        
        let activityViewController = UIActivityViewController(activityItems: [ image ], applicationActivities: nil)
        activityViewController.completionWithItemsHandler = { [weak self] (_, _, _, _) -> Void in
            guard let self = self else {
                return
            }
            self.isPresenting = false
        }
        rootViewController?.present(activityViewController, animated: true, completion: nil)
    }
    
}
